GVFS/GVFS.Common/Maintenance/GitMaintenanceStep.cs

```diff
@@ -143,6 +147,45 @@ protected EventMetadata CreateEventMetadata(Exception e = null)
             return metadata;
         }
 
+        protected bool EnoughTimeBetweenRuns()
+        {
+            if (!this.Context.FileSystem.FileExists(this.LastRunTimeFilePath))
+            {
+                return true;
+            }
+
+            string lastRunTime = this.Context.FileSystem.ReadAllText(this.LastRunTimeFilePath);
+            if (!long.TryParse(lastRunTime, out long result))
+            {
+                this.Context.Tracer.RelatedError("Failed to parse long: {0}", lastRunTime);
+                return false;
+            }
+
+            if (DateTime.UtcNow.Subtract(EpochConverter.FromUnixEpochSeconds(result)) >= this.TimeBetweenRuns)
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        protected void SaveLastRunTimeToFile()
+        {
+            if (!this.Context.FileSystem.TryWriteTempFileAndRename(
+                this.LastRunTimeFilePath,
+                EpochConverter.ToUnixEpochSeconds(DateTime.UtcNow).ToString(),
+                out Exception handledException))
+            {
+                this.Context.Tracer.RelatedError("Failed to record run time: {0}", handledException?.ToString());
+            }
+        }
+
+        protected IEnumerable<int> RunningGitProcessIds()
+        {
+            Process[] allProcesses = Process.GetProcesses();
+            return allProcesses.Where(x => x.ProcessName.Equals("git", StringComparison.OrdinalIgnoreCase)).Select(x => x.Id);

```

Comment:
@derrickstolee can you confirm if 'git.exe' is the only process we care about here?

Reply:
Technically, the `git-<builtin>.exe` pattern exists for historical reasons. We still ship them with Git for Windows.

So yes, it is possible that someone is using these commands directly and we should be extra careful here by watching for `git*.exe`.

Reply:
Good catch! This is definitely an unlikely scenario, but worth guarding against.

Reply:
Spoke with @derrickstolee offline and we'll halt for anything that starts with "git". @alameenshah for reference

Reply:
OK Change of plans!  Too many things start with "git.." (i.e. for example GitServiceDaemon.exe)
We are checking for a weird case here.  And I'll try to explain best I can.

If a git command is running across LooseObjectsMaintenance that is fine.
If a git command is very long running and runs across two calls to LooseObjectsMaintenance, that is not fine.  This is because we may remove a loose object that is a new packfile that the process did not see.  Even if that happens the long running command may not fail.

Given how rare this case is and that the user should be using direct git commands we're sticking with git.exe.

Reply:
Both `gitk` and `git-gui` are written in shell, so they get all of their information by executing `git.exe`.

Here is my understanding of the "worst that can happen" as we are deleting loose objects:

1. A process starts and loads pack-files.
2. We run our step and create a pack-file containing some of our loose objects.
3. We run our step _again_ and start deleting loose objects that we placed into pack-files the last time.
4. The process looks for a loose object, sees that it exists, then tries to read from the file (but it is deleted!). How the process fails at this point is unclear.

Note that if the process starts reading the loose object before we delete it, then the process is fine. If the process looks for existence after we delete it, then we will download the loose object again using the read-object hook.

I think this is an extremely rare error case.

Reply:
@derrickstolee ok thanks for the details!